Python Fundamentals

1. Variables, Expressions and Statements
2. Conditional Expression
3. Functions

Reference: https://www.py4e.com/lessons

### 2.0 Conditional Execution

IF Statement

2.1 Write a program to prompt the user for hours and rate per hour using input to compute gross pay. Pay the hourly rate for the hours up to 40 and 1.5 times the hourly rate for all hours worked above 40 hours. Use 45 hours and a rate of 10.50 per hour to test the program (the pay should be 498.75). You should use input to read a string and float() to convert the string to a number. Do not worry about error checking the user input - assume the user types numbers properly.

In [1]:
hrs = input("Enter Hours:")
type(hrs)
rph = input("Enter rate per hour:")
type(rph)

# convert rate per hour to float
rph= float(rph)
pay = hrs* rph
print("Your pay is: ", pay)

Enter Hours:35
Enter rate per hour:2.75


TypeError: can't multiply sequence by non-int of type 'float'

2.2 Write a program to prompt the user for hours and rate per hour using input to compute gross pay. Use 35 hours and a rate of 2.75 per hour to test the program (the pay should be 96.25). You should use input to read a string and float() to convert the string to a number. Do not worry about error checking or bad user data.

See this [SO post](https://stackoverflow.com/questions/485789/why-do-i-get-typeerror-cant-multiply-sequence-by-non-int-of-type-float/27488582)

In [1]:
hrs = input("Enter Hours:")
print ("Data type is ",type(hrs))
rph = input("Enter rate per hour:")
print ("Data type is ",type(rph))

Enter Hours:10
Data type is  <class 'str'>
Enter rate per hour:2.3
Data type is  <class 'str'>


In [4]:
hrs = float(input("Enter Hours:"))
rph = float(input("Enter rate per hour:"))

# convert rate per hour to float
pay = hrs* rph
print("Your pay is: ", pay)

Enter Hours:35
Enter rate per hour:2.65
Your pay is:  92.75


In [7]:
hrs = float(input("Enter Hours:"))
rph = float(input("Enter rate per hour:"))

# convert rate per hour to float
pay = hrs* rph

if (hrs<=40):
    pay = hrs*rph
if (hrs>40):
    #rph = 1.5*rph
    pay = hrs*1.5*rph
    
print("Your pay is: ", pay)

Enter Hours:40
Enter rate per hour:10.50
Your pay is:  420.0


2.3 Write a program to prompt for a score between `0.0` and `1.0`. If the score is out of range, print an error. If the score is between `0.0` and `1.0`, print a grade using the following table:
    
Score    Grade

`>= 0.9`   A

`>= 0.8`   B

`>= 0.7`   C

`>= 0.6`   D

`< 0.6`    F

If the user enters a value out of range, print a suitable error message and exit. For the test, enter a score of 0.85.

In [1]:
score = float(input("Enter Score: "))

type(score)

if (score>1.0):
    print ("Sorry! Out of range value")
if (score<= 1.0 and score >=0.0):
    if(score >= 0.9):
        print ("A")
    elif(score >= 0.8):
        print ("B")
    elif(score >= 0.7):
        print ("C")
    elif(score >= 0.6):
        print ("D")
    elif(score < 0.6):
        print ("F")
    else:
        print ("Logic is wrong")

Enter Score: 5
Sorry! Out of range value


In [73]:
##### Reference: https://www.hackerrank.com/challenges/py-if-else/problem

N = int(input("Enter a number"))

if (N%2==0 and (N>=2 and N<=5)):
    print ("Not Weird")
elif (N%2==0 and (N>=6 and N<=20)):
    print ("Weird")
elif (N%2==0 and N>20):
    print ("Not Weird")
elif (N%2!=0):
    print ("Weird")
else:
    print ("Incorrect value")

Enter a number-2
Incorrect value


#### Python Division
##### Reference: https://www.hackerrank.com/challenges/python-division/problem?h_r=next-challenge&h_v=zen

Task 
Read two integers and print two lines. The first line should contain integer division,  // . The second line should contain float division,  / .

You don't need to perform any rounding or formatting operations.

In [78]:
N = int(input("Enter a number: "))
Y = int(input("Enter a number: "))

print (N//Y)
Y = float(Y)
print (N/Y)

Enter a number: 5
Enter a number: 4
1
1.25


#### Python Loops

Loops are control structures that iterate over a range to perform a certain task.

There are two kinds of loops in Python.

A **for** loop:

```{python}

for i in range(0, 5):
    print i
```
And a **while** loop:

```{python}
i = 0
while i < 5:
    print i
    i += 1
```


Task (Reference: https://www.hackerrank.com/challenges/python-loops/problem?h_r=next-challenge&h_v=zen)

Read an integer `N`. For all non-negative integers `i<N`, print `i2 (i squared)`. See the sample for details.

Input Format

The first and only line contains the integer, .

Constraints
1<=N<=20

Output Format

Print `N`lines, one corresponding to each `i`.

In [81]:
# Solution

n = int(input())

if (n<=20):
    for i in range (0,n):
        print (i*i)
else:
    print("Out of range")

5
0
1
4
9
16


### 3.0 Functions

A function begins with the keyword, `def` followed by the function name and parenthesis. Example, `def calculateMarks(){}`. When a function is created, the variables inside the parenthesis are called `Parameters`. When the function is called/invoked, the variables passed into the function are called `arguments`. A function is written to return a value. The functions returning value are called as `Fruitful` functions. Example, `def calculateMarks(){return 1}`. A function not returning a value are called as `void` functions.

#### 3.1 Exercise

Rewrite your pay computation with time-and-a-half for overtime and create a function called **computePay** which takes two parameters `(hours and rate)`.

Enter Hours: 45

Enter Rate: 10

Pay: 475.0

Reference: https://www.youtube.com/watch?v=l93PhBUJ_S0&feature=youtu.be

In [58]:
# Answer
def computePay (hours, rate):
    print ("In computePay", hours, rate)
    
    if (hours > 40):
        reg = rate * hours
        otp = (hours-40.0) * (rate*0.5)
        pay = reg + otp
    else:
        pay = hours * rate
        print ("Returning ", pay)
        return pay
    
#     sh = input("Enter hours: ")
#     sr = input("Enter rate: ")
#     fh = float(sh)
#     fr = float(sr)
    
#     xp = computePay(fh, fr)
#     print ("Pay: ", xp)

#     # convert rate per hour to float
#     pay = hours* rate
    
#     if (hours<=40):
#         pay = hours*rate
#     elif (hours>40):
#         #rph = 1.5*rph
#         pay = hours*1.5*rate
#     else:
#         print ("\n Invalid number of hours or rate")
#         return 0
#         #break
       
#     return pay


In [59]:
print ("My pay is: ", computePay(10,15))

In computePay 10 15
Returning  150
My pay is:  150


In [60]:
print ("My pay is: ", computePay(40,10))

In computePay 40 10
Returning  400
My pay is:  400


In [61]:
print ("My pay is: ", computePay(40,-1))

In computePay 40 -1
Returning  -40
My pay is:  -40


#### 3.2 Exercise

Rewrite the grade program from the previous chapter using a function called **computeGrade** that takes a score as its parameter and returns a grade as a string.

Score   Grade

`>= 0.9`     A

`>= 0.8`     B

`>= 0.7`     C

`>= 0.6`     D

`< 0.6`      F

Enter score: 0.95

A

Enter score: perfect

Bad score

Enter score: 10.0

Bad score

Enter score: 0.75

C

Enter score: 0.5

F

Run the program repeatedly to test the various different values for input.

In [1]:
# Answer

def computeGrade (score):
    if (score >= 0.9 and score<= 1.0):
        return "A"
    elif (score >= 0.8 and score <0.9):
        return "B"
    elif (score >= 0.7 and score <0.8):
        return "C"
    elif (score >= 0.6 and score <0.7):
        return "D"
    elif (score < 0.6 ):
        return "F"
    elif (type(score)=='str'):
        return "Bad score" 

In [2]:
score = float(input("Enter score:"))
computeGrade(score)

Enter score:0.9


'A'

In [4]:
score = float(input("Enter score:"))
computeGrade(score)

Enter score:0.8


'B'

In [42]:
score = float(input("Enter score:"))
computeGrade(score)

Enter score:0.7


'C'

In [45]:
score = float(input("Enter score:"))
computeGrade(score)

Enter score:0.2


'F'

In [5]:
score = input("Enter score:")
computeGrade(score)

Enter score:p


TypeError: '>=' not supported between instances of 'str' and 'float'